### ISSUE TITLE
Cannot reliably test against different IP addresses

### ISSUE THREAD

**Issue by letitz** [web-platform-tests/wpt#26166]
Hi there, I believe this issue deserves the `type:untestable` tag but I am not empowered to tag it myself.

[CORS-RFC1918](https://wicg.github.io/cors-rfc1918/) is a spec aiming to prevent malicious public websites (e.g. `http://evil.com`) from attacking users' devices on the local network (e.g. smart fridge) or companies' intranet services.

It does so by partitioning IP addresses into 3 address spaces, ordered from most to least private:

* local: contains only localhost
* private: contains IPv4 addresses designated private by RFC1918 and non-publicly routable IPv6 addresses
* public: all other addresses

See https://github.com/web-platform-tests/wpt/pull/26135 for an initial pass at introducing tests for this spec. As @mikewest pointed out in [the review](https://crrev.com/c/2475875), it is currently impossible for tests to assert the address space from which they are loaded: on my machine they are served from localhost, but that is certainly not the case for https://wpt.live.

In full generality, testing this spec properly would require being able to load a web page from the local, private or public address spaces at will. This could be achieved by wptserve actually serving the data from localhost, a private or a public IP, or by fooling the test browser into believing that somehow.

**Comment by stephenmcgruer** [comment]
cc @web-platform-tests/wpt-core-team

**Comment by Hexcles** [comment]
With some rare exceptions (e.g. mobile testing), tests are almost always served from localhost in an automated system (wpt.live is for manual testing only).

To support private and public addresses from the infra side, we could bind additional IPs to a loopback device, which of course requires root permissions. I think we can do this on Taskcluster and Azure Pipelines, but not sure whether it's possible in each browser vendor's CI. There's also a slight complication here regarding the public IP -- we'd probably need to buy one to make sure it's never used for anything real.

To redirect IP addresses to localhost from the browser side, we'd need to change each browser.

In short, I think both approaches require an RFC to coordinate with browser vendors.

**Comment by stephenmcgruer** [comment]
Forgive my networking ignorance, but are the domains we bind to in `/etc/hosts` treated as localhost by the browser (aka `local` scope) in these scenarios? See the output of `./wpt make-hosts-file`:

```
127.0.0.1       op73.not-web-platform.test
127.0.0.1       xn--n8j6ds53lwwkrqhv28a.www.not-web-platform.test
127.0.0.1       op88.web-platform.test
127.0.0.1       xn--n8j6ds53lwwkrqhv28a.www1.not-web-platform.test
127.0.0.1       op35.not-web-platform.test
127.0.0.1       op29.web-platform.test
...
```

If someone were to set a URL as those (via [server-side substitution](https://web-platform-tests.org/writing-tests/server-pipes.html#sub), e.g. using `{{hosts[alt][]}}` I think?), would it count as `public` or `local`?

**Comment by mikewest** [comment]
@stephenmcgruer: Yes. Anything bound to loopback (127.0.0.0/8) is considered to be a "local" endpoint, regardless of name. We look at the IP address associated with the socket, not at the domain.

@Hexcles: An alternative could be to teach the test driver some new tricks to classify some IP addresses as local and others as public. That would put the burden on the browsers to do the categorization in some reasonable way, but it wouldn't be out of the question.

**Comment by stephenmcgruer** [comment]
I would personally lean towards a webdriver extension, since 'buying a public IP address' seems future-brittle, but in either case @Hexcles is correct and the next step is likely [an RFC](https://github.com/web-platform-tests/rfcs). Please feel free to pick your favorite solution to the problem and write up a short RFC with summary/details/risk sections, then browser vendors/etc can chime in :)

**Comment by letitz** [comment]
Alright, thanks! I'll read up on webdriver and how we could leverage that, and come back with an RFC in hand :)

**Comment by foolip** [comment]
#72 was merged, considering this fixed.
